### PR TITLE
pkgconfig: Allow passing Dependency objects to library(_private)

### LINKED ISSUE #2725: Extract compiler/linker flags from threads dependency
GLib can build against different threading implementation, depending on the platform, as it provides portable threading primitives. In order to properly communicate the dependency to projects depending on GLib itself, the compiler and linker flags for the threading implementation are passed inside the glib-2.0 pkgconfig file:

```
Name: GThread
Description: Thread support for GLib
Requires: glib-2.0
Version: @VERSION@
Libs: -L${libdir} -lgthread-2.0 @G_THREAD_LIBS@
Cflags: @G_THREAD_CFLAGS@
```

When building GLib with Meson, the `G_THREAD_LIBS` and `G_THREAD_CFLAGS` are empty, as GLib uses `dependency("threads")`. This means that projects using GLib and the threading primitives are now missing the appropriate compiler and linker flags depending on the platform, and they'll have to manually add the dependency on the threading library on their own build system, and hope it's the same library used by GLib.

A possible solution would be to allow calling `get_pkgconfig_variable("Cflags")` and `get_pkgconfig_variable("Libs")` on the "threads" dependency object.

### PR DESCRIPTION
**Pull request by xclaesse**
Special case ThreadDependency by taking compiler's flags and
PkgConfigDependency by adding them in requires(.private) instead. For
other Dependency objects just take their link_args.

Closes #2725

### CODE REVIEW

**Comment by nirbheek** [comment]
This requires a test case, preferably something that tests with as many dependency types as possible. f.ex., this is definitely broken with dependency types that don't yield `-lfoo` in link_args. For those, you'll have to split it into `-Lbar -lfoo` manually.

**Comment by xclaesse** [comment]
> dependency types that don't yield -lfoo

Any example?

**Comment by nirbheek** [comment]
> > dependency types that don't yield -lfoo
> 
> Any example?

BoostDependency when not using pkg-config files will do this.

... actually, now that I think of it, a lot of dependency files use `PkgConfigDependency` internally. F.ex: [GLDependency](https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/ui.py#L42), BoostDependency, SDL2Dependency, etc. This is a bit problematic.

I wonder if we can set `self` in these cases and replace the actual object or something like that.

**Comment by xclaesse** [comment]
I see, they should set self.pcdep so we can use hasattr() on it.

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -617,6 +618,7 @@ def __init__(self, environment, kwargs):
                     self.link_args = self.pkgdep.get_link_args()
                     self.version = self.pkgdep.get_version()
                     self.is_found = True
+                    self.pcdep = pkgdep
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

undefined name 'pkgdep'

File: mesonbuild/modules/pkgconfig.py [hunk 2]
```diff
@@ -133,25 +136,36 @@ def generate_libs_flags(libs):
...
         processed_libs = []
+        processed_reqs = []
         for l in libs:
             if hasattr(l, 'held_object'):
                 l = l.held_object
-            if not isinstance(l, (build.SharedLibrary, build.StaticLibrary, str)):
-                raise mesonlib.MesonException('Library argument not a library object nor a string.')
-            processed_libs.append(l)
-        return processed_libs
+            if hasattr(l, 'pcdep')
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__

SyntaxError: invalid syntax

File: mesonbuild/modules/pkgconfig.py [hunk 3]
```diff
@@ -133,25 +136,36 @@ def generate_libs_flags(libs):
...
+                processed_reqs += l.pcdep
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__

unexpected indentation

File: mesonbuild/modules/pkgconfig.py [hunk 4]
```diff
@@ -16,12 +16,76 @@
...
+        self.cflags += mesonlib.stringlistify(cflags)
+
+    def _process_libs(self, libs, public):
+        libs = mesonlib.listify(libs)
+        processed_libs = []
+        processed_reqs = []
+        processed_cflags = []
+        for l in libs:
+            if hasattr(l, 'held_object'):
+                l = l.held_object
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__

ambiguous variable name 'l'

File: mesonbuild/modules/pkgconfig.py [hunk 5]
```diff
@@ -16,12 +16,76 @@
...
+                processed_libs += l.get_link_args()
+                processed_cflags += l.get_compile_args()
+            elif isinstance(l, (build.SharedLibrary, build.StaticLibrary)):
+                processed_libs.append(l)
+                if public:
+                    self.add_priv_libs(l.get_dependencies())
+            elif isinstance(l, str):
+                processed_libs.append(l)
+            else:
+                raise mesonlib.MesonException('%s argument not a string, library or dependency object.' % field)
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__

undefined name 'field'

**Comment by jpakkane** [comment]
Please move the test to `run_unittests.py`. That is the correct place for any tests that do complex system checking. Project tests are designed for simple run & go tests. Furthermore we try to keep tests running pkg-config away from common tests so that they can be run everywhere without thinking (there may be some others, we should fix them).

**Comment by nirbheek** [comment]
> Furthermore we try to keep tests running pkg-config away from common tests so that they can be run everywhere without thinking (there may be some others, we should fix them).

If you move this to unit tests, please put it in `AllPlatformTests` and skip the test if pkg-config is not found. We already do this for other tests, so you can copy that code from there.

We should ideally move all project tests that use pkg-config from linuxlike to common tests and skip them if pkg-config is not found because even though we shouldn't require pkg-config for tests on Windows and macOS, we still use pkg-config on those platforms if it's available.

File: mesonbuild/modules/pkgconfig.py [hunk 6]
```diff
@@ -16,12 +16,87 @@
...
 
 from .. import build
+from .. import dependencies
 from .. import mesonlib
 from .. import mlog
 from . import ModuleReturnValue
 from . import ExtensionModule
 from ..interpreterbase import permittedKwargs
 
+class DepsHolder:
```
Comment (jpakkane) [inline mesonbuild/modules/pkgconfig.py]: Don't call this a holder. It has a very specific meaning in the context of Meson. A holder is always an InterpreterObject that exposes something to the user. Internal helper classes should be differentiated by having a different name.

**Pull request merged**